### PR TITLE
GAPIR crash reporting

### PR DESCRIPTION
**Pull request by baldwinn860**
Now crash dumps in GAPIR will get uploaded to GAPIS via the server connection such that GAPIS can do the final upload to the crash server. Fixes the GAPIR crash reporting sub-task of #407.

### CODE REVIEW

File: core/app/crash/crash.go [hunk 1]
```diff
@@ -19,20 +19,33 @@
 package crash
```
Comment (ben-clayton) [inline core/app/crash/crash.go]: Please don't change this package. Just add another method to the `crash/reporting` package to deal with external process crashes.

File: core/app/analytics/crash_reporter.go [hunk 2]
```diff
@@ -33,11 +33,13 @@ func init() {
 	crash.Register(onCrash)
 }
 
-func onCrash(e interface{}, s stacktrace.Callstack) {
+func onCrash(r crash.Report, appName, appVersion string) {
```
Comment (ben-clayton) [inline core/app/analytics/crash_reporter.go]: Please revert this change. See my other comment.

File: core/app/crash/reporting/encoder.go [hunk 3]
```diff
@@ -18,48 +18,60 @@ import (
 	"bytes"
 	"io"
 	"mime/multipart"
+
+	"github.com/google/gapid/core/app/crash"
```
Comment (ben-clayton) [inline core/app/crash/reporting/encoder.go]: I'm not convinced that using the `crash` package here is better than no intra-dependencies.

File: core/app/crash/reporting/encoder_test.go [hunk 4]
```diff
@@ -44,19 +41,6 @@ func TestEncoder(t *testing.T) {
 				"Content-Disposition: form-data; name=\"version\"\r\n" +
 				"\r\n" +
 				"TestApp:V1.2.3\r\n" +
```
Comment (ben-clayton) [inline core/app/crash/reporting/encoder_test.go]: And now we're no longer testing a bunch of logic. Please revert.

File: core/app/crash/reporting/reporting.go [hunk 5]
```diff
@@ -47,18 +48,23 @@ var (
...
+					}
 				}
 			}
-			err := reporter{appName, appVersion, osName, osVersion}.report(s, crashURL)
+			rep := reporter{app, version}
+			response, err := rep.report(r, crashURL)
 			if err != nil {
 				log.E(ctx, "%v", err)
+			} else {
+				log.I(ctx, "%s", response)
```
Comment (ben-clayton) [inline core/app/crash/reporting/reporting.go]: This for debugging?

File: gapis/replay/executor/executor.go [hunk 6]
```diff
@@ -48,12 +51,14 @@ func Execute(
 	payload protocol.Payload,
 	decoder builder.ResponseDecoder,
 	connection io.ReadWriteCloser,
-	memoryLayout *device.MemoryLayout) error {
+	memoryLayout *device.MemoryLayout,
```
Comment (ben-clayton) [inline gapis/replay/executor/executor.go]: Just use the `device.Configuration`, it contains both the `MemoryLayout` and `OS`.

File: core/app/crash/crash.go (hunk 1 above)
Comment (baldwinn860) [reply core/app/crash/crash.go]: What about the shutdown function setter? Without this change the cancelFunc set in app.handleCrashSignals could preempt any other crash handlers registered and shut down the app before they can report anything.

Comment (ben-clayton) [reply core/app/crash/crash.go]: > Without this change the cancelFunc set in app.handleCrashSignals could preempt any other crash handlers registered and shut down the app before they can report anything.

Nothing in the crash handler should depend on the context cancellation state. Are you seeing this somewhere?

File: gapis/replay/executor/executor.go (hunk 6 above)
Comment (baldwinn860) [reply gapis/replay/executor/executor.go]: device.Configuration will give us the memoryLayout for all ABIs this device supports, but we need the specific one that the capture was taken with.

File: core/app/crash/reporting/reporting.go [hunk 7]
```diff
@@ -74,28 +80,25 @@ func Disable() {
 	}
 }
 
-type reporter struct {
-	appName    string
-	appVersion string
-	osName     string
-	osVersion  string
+func ReportMinidump(r Reporter, minidumpName string, minidumpData []byte) error {
```
Comment (ben-clayton) [inline core/app/crash/reporting/reporting.go]: Godocs please

File: core/app/crash/reporting/reporting.go [hunk 8]
```diff
@@ -74,28 +80,25 @@ func Disable() {
...
+		if err := r.reportMinidump(minidumpName, minidumpData, crashURL); err != nil {
+			return err
+		}
+	}
+	return nil
 }
 
-func (r reporter) report(s stacktrace.Callstack, endpoint string) error {
-	stacktrace := s.String()
+type Reporter struct {
```
Comment (ben-clayton) [inline core/app/crash/reporting/reporting.go]: Godocs please

File: core/cc/crash_handler.cpp [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+void CrashHandler::unsetHandlerFunction() {
+    mHandlerFunction = defaultHandlerFunction;
+}
+
+bool CrashHandler::handleMinidump(const std::string& minidumpPath, bool succeeded) {
+    return mHandlerFunction(minidumpPath, succeeded);
+}
+
+
+}
```
Comment (ben-clayton) [inline core/cc/crash_handler.cpp]: `}  // namespace core`

File: core/cc/file_reader.cpp [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+        GAPID_ERROR("Failed to get size of file")
+        rewind(mFile);
+        return 0u;
+    }
+    uint64_t size = static_cast<uint64_t>(sizeTemp);
+    rewind(mFile);
+    return size;
+}
+
+}
```
Comment (ben-clayton) [inline core/cc/file_reader.cpp]: `}  // namespace core`

File: core/cc/file_reader.h [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+#define CORE_FILE_READER_H
+
+#include "stream_reader.h"
+
+#include <stdio.h>
+#include <stdint.h>
+
+namespace core {
+
+// FileReader is a pure-virtual interface used to read from binary files.
```
Comment (ben-clayton) [inline core/cc/file_reader.h]: Doesn't look very pure-virtual interface like to me!

File: gapir/cc/crash_uploader.cpp [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+        return succeeded;
+    });
+}
+
+CrashUploader::~CrashUploader() {
+    GAPID_INFO("destroying uploader");
+    mCrashHandler.unsetHandlerFunction();
+}
+
+}
```
Comment (ben-clayton) [inline gapir/cc/crash_uploader.cpp]: `}  // namespace gapir`

File: gapir/cc/crash_uploader.h [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+  CrashUploader(core::CrashHandler& crash_handler, const ServerConnection& conn);
+  ~CrashUploader();
+
+private:
+
+  core::CrashHandler& mCrashHandler;
+  const ServerConnection& mConnection;
+};
+
+}
```
Comment (ben-clayton) [inline gapir/cc/crash_uploader.h]: `}  // namespace gapir`

File: gapir/cc/server_connection.cpp [hunk 14]
```diff
@@ -126,4 +126,31 @@ bool ServerConnection::post(const void* postData, uint32_t postSize) const {
     return true;
 }
 
+bool ServerConnection::crash(const std::string& filename, const void* crashData, uint32_t crashSize) const {
+    GAPID_DEBUG("CRASH: [%s] %p (%d)", filename.c_str(), crashData, crashSize);
```
Comment (ben-clayton) [inline gapir/cc/server_connection.cpp]: GAPID_WARNING?

File: gapir/cc/server_connection.h [hunk 15]
```diff
@@ -56,11 +56,16 @@ class ServerConnection {
     // the posting was successful false otherwise.
     bool post(const void* postData, uint32_t postSize) const;
 
+    // Post a crash minidump from the given address with the given filename to the server. Returns
+    // true if the posting was successful false otherwise.
+    bool crash(const std::string& filename, const void* crashData, uint32_t crashSize) const;
```
Comment (ben-clayton) [inline gapir/cc/server_connection.h]: maybe rename `postCrashdump`? The other methods are verbs, and this doesn't sound like a function you'd want to call. 😄
